zkevm-circuits/src/mpt_circuit.rs

```diff
@@ -205,21 +238,43 @@ impl<F: Field> MPTConfig<F> {
                     ifx! {not!(f!(q_not_first)) => {
                         require!(a!(is_start) => true);
                     }};
+                    // When q_row > 0, we're in the middle of some node rows, all flags needs to be 0 
+                    // Otherwise q_row == 0, we start at a new node and corresponding flag needs to be 1
+                    // taken care by  _ => require!(true => false)
+                    ifx! {a!(q_row)  => {
+                        require! ((a!(is_start) + a!(is_branch) + a!(is_account) + a!(is_storage)) => 0.expr());
+                    }} 
                     // Main state machine
                     matchx! {
                         a!(is_start) => {
+                            require!(a!(q_row) + a!(q_node) => 0.expr());                            
                             state_machine.start_config = StartConfig::configure(meta, &mut cb, ctx.clone());
                         },
                         a!(is_branch) => {
+                            // q_node[cur] - count == q_node[cur-count]
+                            // Start -> Branch || Branch -> Branch
+                            let diff1 = a!(q_node) - (ExtensionBranchRowType::Count as i32).expr() - a!(q_node, -(ExtensionBranchRowType::Count as i32));
```

Comment:
I'm a bit surprised this rotation works to be honest. The difficulty with rotation to previous rows is that we have to check that the rotation on the current row doesn't get to some row < 0. This actually wraps around the circuit (so a rotation of -1 on the first row accesses the data on the last row), but the additional difficulty is that on the last rows of the circuit there's random data stored to enable zk. A branch is almost always used after the initial start which uses 2 rows, and `ExtensionBranchRowType::Count` is a pretty large value, so this seems to read from this garbage data (halo2 calls these poisoned cells, and the rows containing the random data unusable rows).

So most of the time you actually want to avoid using previous values because they are trickier, it's easier (and basically equivalent) to only look to the next rows instead.

Reply:
`The difficulty with rotation to previous rows is that we have to check that the rotation on the current row doesn't get to some row < 0. This actually wraps around the circuit (so a rotation of -1 on the first row accesses the data on the last row), but the additional difficulty is that on the last rows of the circuit there's random data stored to enable zk.` 
For the specific arrangement dealing with rotation, does halo2 have docs for it?

Reply:
Not really any good ones that I know of, they do touch on it in a couple of places, for example:
- https://zcash.github.io/halo2/design/proving-system/lookup.html?highlight=blinding#zero-knowledge-adjustment
- https://github.com/privacy-scaling-explorations/halo2/blob/fcdd5b910d92ab278d679cf8aa176c885455aa6c/halo2_proofs/src/plonk/circuit.rs#L2011
- https://github.com/privacy-scaling-explorations/halo2/blob/fcdd5b910d92ab278d679cf8aa176c885455aa6c/halo2_proofs/src/dev.rs#L636 (this it where the cell values are set as "poisoned" in the MockProver so it will return an error).